Tests/UpstreamIntegrationTests/util/XCTestCase+AnyCodableAsserts.swift

```diff
@@ -0,0 +1,729 @@
+                current[key] = merge(current: currentValue, new: newValue)
+            default:
+                if current[key] is String {
+                    continue
+                }
+                current[key] = newValue
+            }
+        }
+        return current
+    }
+
+    /// Constructs a key path dictionary from a given key path component array, and the final value is
+    /// assigned the original path string used to construct the path
+    private func construct(path: [String], pathString: String) -> [String: Any] {
+        var path = path
+        let first = path.removeFirst()
+        let result: [String: Any]
+        if path.isEmpty {
+            result = [first: pathString]
+            return result
+        } else {
+            return [first: construct(path: path, pathString: pathString)]
+        }
+    }
+
+    private func generatePathTree(paths: [String]) -> [String: Any]? {
+        // Matches array subscripts and all the inner content. Captures the surrounding brackets and inner content: ex: "[123]", "[*123]"
+        let arrayIndexRegex = #"(\[.*?\])"#
+        // Matches key path access in the style of: "key0\.key1.key2[1][2].key3". Captures each of the groups separated by `.` character and ignores `\.` as nesting.
+        // the path example would result in: ["key0\.key1", "key2[1][2]", "key3"]
+        let jsonNestingRegex = #"(.+?)(?<!\\)(?:\.|$)"#
+        var tree: [String: Any] = [:]
+
+        for exactValuePath in paths {
+            var allPathComponents: [String] = []
+            var pathExtractionSuccessful: Bool = true
+
+            // Break the path string into its component parts
+            let splitByNesting = getCapturedRegexGroups(text: exactValuePath, regexPattern: jsonNestingRegex)
+            for pathComponent in splitByNesting {
+                guard let validComponent = pathComponent else {
+                    XCTFail(#"""
+                        TEST ERROR: unable to extract valid key path component from path: \#(exactValuePath)
+                        Skipping this path in validation process.
+                    """#)
+                    pathExtractionSuccessful = false
+                    break
+                }
+                let pathComponent = validComponent.replacingOccurrences(of: "\\", with: "")
+
+                // Get all array access levels for the given pathComponent, if any
+                // KNOWN LIMITATION: this regex only extracts all open+close square brackets and inner content ("[___]") regardless
+                // of their relative position within the path component, ex: "key0[2]key1[3]" will be interpreted as: "key0" with array component "[2][3]"
+                let arrayComponents = getCapturedRegexGroups(text: pathComponent, regexPattern: arrayIndexRegex)
+
+                // If array components are detected, extract just the path component before array components if it exists
+                if !arrayComponents.isEmpty {
```

Comment:
code style: For readability it's usually better to use the reversed if (something) else
instead of if (!something) else (aka something) ..

Reply:
don't forget to reverse the condition `if (arrayComponents.isEmpty)`